epi/epi_functions.py

```diff
@@ -0,0 +1,265 @@
+pd.options.display.max_colwidth = 50
+pd.options.display.max_columns = 40
+
+
+def dump_epi_tb4(
+    datum=None, *, original_csv=None, new_csv="latest_dump_tb4.csv", impute=True
+):
+    """Dump data from source_csv into a rainbow of cumsum columns.
+
+    A Pandas DataFrame and a side effect of two additional files in the folder
+    where the script was called from (result_csv, log showing differences from
+    source_csv.
+
+    :param datum: Datum in ISO format (i.e. 2020-10-26).
+    :param original_csv: A csv file that has been created in the previous run of the
+      script. This file may not exist and it's fine, just leave it blank.
+    :param new_csv: Name of the resulting file. Default is "latest_dump_tb4.csv,
+      but can be anything, really.
+    :param impute: Whether to ask about imputing missing dates or not. Default is True.
+    """
+
+    # Algorithm to give you the general idea of what this function is suppose to do:
+    #
+    # inputs:
+    #   <date>
+    # implicit inputs:
+    #   <original_csv>
+    #   <xlsx>
+    #
+    # 1: <xlsx> checks
+    #   - based on date provided, find corresponding <xlsx> file. If missing yield
+    #     informative error if file is not found (directions on how to structure the
+    #     filename)
+    #   - check that tb4 sheet exists with informative message on how to format the
+    #     sheet name
+    # 2: data integrity check
+    #   - expect first column to be date and monotinically increasing/decreasing
+    #   - expect columns 2:11 to be counts for age groups 0-85+ for men
+    #   - expect columns 13:22 to be counts for age groups 0-85+ for women
+    #   - all values for columns 2:22 should be >= 0 or NA (empty)
+    # 3: missing dates
+    #   - if any missing dates, ask user if assuming NA is ok
+    #     - if ok, add missing dates NA for all columns
+    #     - if not, abort
+    # 4: calculate sums
+    #   - calculate cumulative sums per age group per gender (20 new variables)
+    #   - calculate row-wise age group sums (irrespective of gender) of cumulative sums
+    #     (10 new variables)
+    #   - calculate row-wise sums for each gender of cumulative sums (2 new variables)
+    #   - calculate grand row-sums for genders combined of cumulative sums (1 new
+    #     variable)
+    #
+    # output of the script is these new 34 variables (<new_csv>) and should not
+    # overwrite <original_csv>
+    #
+    # 5: diff
+    #   - create a log of differences between <original_csv> and new table created in 4
+    #   - create a heatmap showing potential differences between <original_csv> and
+    #     <new_csv>
+
+    # original_csv = "original_tb4.csv"
+    # new_csv = "latest_dump_tb4.csv"
+    # datum = "2020-10-22"
+
+    if not datum:
+        raise ValueError("You did not enter a date.")
+
+    input_datum = datetime.datetime.strptime(datum, "%Y-%m-%d").strftime("%d%m%Y")
+
+    xlsx = glob.glob(pathname=f"*{input_datum}.xlsx")
```

Comment:
Would it be too heretic/problematic to take the latest date from glob if date is missing?

Reply:
If it turns out that typing the date is too cumbersome, let's add this feature too.